File: typescript/src/lib/reconcile.ts
import { PagouHttpClient } from "./http.js";
import { NotFoundError } from "./errors.js";
import { TERMINAL_PAID_STATUSES, type Transaction, type TransactionStatus } from "./types.js";

export type FulfillmentDecision = "fulfill" | "wait" | "cancel";

// Terminal failure/cancel states: stop waiting, release the order.
const TERMINAL_FAILED: ReadonlySet<TransactionStatus> = new Set([
  "canceled",
  "expired",
  "refused",
]);

/** Maps a transaction status to a business decision. Never fulfill on a pending state. */
export function decideFulfillment(status: TransactionStatus): FulfillmentDecision {
  if (TERMINAL_PAID_STATUSES.has(status)) return "fulfill";
  if (TERMINAL_FAILED.has(status)) return "cancel";
  return "wait";
}

/**
 * Server-side reconciliation: fetch the transaction from the API (the source of
 * truth) and decide whether to fulfill. Business state changes only on this
 * confirmed result, never on an unverified webhook body.
 */
export async function reconcileTransaction(
  id: string,
  client = new PagouHttpClient(),
): Promise<{ transaction: Transaction; decision: FulfillmentDecision } | null> {
  try {
    const { data: transaction } = await client.requestData<Transaction>({
      method: "GET",
      path: `/v2/transactions/${id}`,
    });
    return { transaction, decision: decideFulfillment(transaction.status) };
  } catch (error) {
    if (error instanceof NotFoundError) return null;
    throw error;
  }
}

File: go/internal/pagou/reconcile.go
package pagou

import "fmt"

// FulfillmentDecision is the business outcome of reconciling a transaction.
type FulfillmentDecision string

const (
	Fulfill FulfillmentDecision = "fulfill"
	Wait    FulfillmentDecision = "wait"
	Cancel  FulfillmentDecision = "cancel"
)

// terminalFailed are the terminal failure/cancel states: stop waiting, release
// the order.
var terminalFailed = map[TransactionStatus]struct{}{
	"canceled": {}, "expired": {}, "refused": {},
}

// DecideFulfillment maps a transaction status to a business decision. It never
// fulfills on a pending state.
func DecideFulfillment(status TransactionStatus) FulfillmentDecision {
	if _, ok := terminalPaidStatuses[status]; ok {
		return Fulfill
	}
	if _, ok := terminalFailed[status]; ok {
		return Cancel
	}
	return Wait
}

// Reconciliation is the outcome of a server-side reconcile.
type Reconciliation struct {
	Transaction Transaction
	Decision    FulfillmentDecision
}

// ReconcileTransaction fetches the transaction from the API (the source of
// truth) and decides whether to fulfill. Business state changes only on this
// confirmed result, never on an unverified webhook body. A missing transaction
// returns (nil, nil).
func ReconcileTransaction(client *Client, id string) (*Reconciliation, error) {
	res, err := RequestData[Transaction](client, RequestParams{
		Method: "GET",
		Path:   fmt.Sprintf("/v2/transactions/%s", id),
	})
	if err != nil {
		if IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	return &Reconciliation{Transaction: res.Data, Decision: DecideFulfillment(res.Data.Status)}, nil
}
